### PR TITLE
Optimize foreachPar_: can handle large input fast #1182

### PR DESCRIPTION
**Pull request by evis**
Parallelize effects manually, using fibers.

Benchmark results on input data with 10000 elements:

```
[info] Benchmark                                 Mode  Cnt   Score   Error  Units
[info] ForeachParN_Benchmark.foreachParN_       thrpt    5  17.603 ± 0.584  ops/s
[info] ForeachParN_Benchmark.naiveForeachParN_  thrpt    5   0.056 ± 0.005  ops/s
```

### CODE REVIEW

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -2242,12 +2242,36 @@ object ZIO {
...
+  def foreachPar_[R, E, A](as: Iterable[A])(f: A => ZIO[R, E, Any]): ZIO[R, E, Unit] = {
+    val size = as.size
+    for {
+      parentId <- ZIO.fiberId
+      result   <- Promise.make[E, Unit]
+      succeed  <- Ref.make(0)
+      _ <- ZIO.traverse_(as) {
+            f(_).fork >>= { fiber =>
+              result.await.catchAll(_ => fiber.interruptAs(parentId)).fork *>
+                fiber.join.foldM(result.fail, _ => {
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/zio/ZIO.scala]: This will block until the fiber is done, effectively running each effect in sequence.

**Comment by evis** [comment]
Benchmark results with bugfix are even better:

```
[info] Benchmark                                 Mode  Cnt   Score   Error  Units
[info] ForeachParN_Benchmark.foreachParN_       thrpt    5  27.325 ± 0.386  ops/s
```

**Comment by iravid** [comment]
@evis The benchmark is for `foreachPar_`, right? Could you rename it and the methods in it?

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 2]
```diff
@@ -2242,12 +2242,37 @@ object ZIO {
...
-      .refailWithTrace
+  def foreachPar_[R, E, A](as: Iterable[A])(f: A => ZIO[R, E, Any]): ZIO[R, E, Unit] = {
+    val size = as.size
+    for {
+      parentId <- ZIO.fiberId
+      result   <- Promise.make[E, Unit]
+      succeed  <- Ref.make(0)
+      _ <- ZIO.traverse_(as) {
+            f(_)
+              .foldM(result.fail, _ => {
```
Comment (iravid) [inline core/shared/src/main/scala/zio/ZIO.scala]: Currently defects in a fiber will not cause the other fibers to be terminated. So let's do this:
```suggestion
              .foldCauseM(result.halt, _ => {
```

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 3]
```diff
@@ -2242,12 +2242,37 @@ object ZIO {
...
+                succeed.update(_ + 1) >>= { succeed =>
+                  ZIO.when(succeed == size)(result.succeed(()))
+                }
+              })
+              .fork >>= { fiber =>
+              result.await.catchAll(_ => fiber.interruptAs(parentId)).fork
```
Comment (iravid) [inline core/shared/src/main/scala/zio/ZIO.scala]: And here we need `catchAllCause`

**Comment by regiskuckaertz** [comment]
wow 😮 massive improvement 💪 🕺

**Comment by adamgfraser** [comment]
Fantastic!

**Pull request merged**